File: SPRING_BOOT/Modern_Hospital_Management_System/src/main/java/emranhss/com/Modern_Hospital_Management_System/service/AmbulanceService.java
package emranhss.com.Modern_Hospital_Management_System.service;

import emranhss.com.Modern_Hospital_Management_System.dto.request.AmbulanceRequest;
import emranhss.com.Modern_Hospital_Management_System.dto.response.AmbulanceResponse;

import java.util.List;

public interface AmbulanceService {

    AmbulanceResponse create(AmbulanceRequest request);

    AmbulanceResponse getById(Long id);

    List<AmbulanceResponse> getAll();

    AmbulanceResponse update(Long id, AmbulanceRequest request);

    AmbulanceResponse updateStatus(Long id, String status);

    void delete(Long id);

    List<AmbulanceResponse> getByStatus(String status);

    Long getAvailableCount();
}

File: SPRING_BOOT/Hospital_Management_system_angular/src/app/services/emergency/ambulance.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Ambulance } from '../../models/emergency';

@Injectable({ providedIn: 'root' })
export class AmbulanceService {
  private api = environment.apiUrl + 'emergency/ambulances';

  constructor(private http: HttpClient) {}

  create(ambulance: Ambulance): Observable<Ambulance> {
    return this.http.post<Ambulance>(this.api, ambulance);
  }

  getById(id: number): Observable<Ambulance> {
    return this.http.get<Ambulance>(`${this.api}/${id}`);
  }

  getAll(): Observable<Ambulance[]> {
    return this.http.get<Ambulance[]>(this.api);
  }

  update(id: number, ambulance: Ambulance): Observable<Ambulance> {
    return this.http.put<Ambulance>(`${this.api}/${id}`, ambulance);
  }

  updateStatus(id: number, status: string): Observable<Ambulance> {
    return this.http.put<Ambulance>(`${this.api}/${id}/status`, { status });
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.api}/${id}`);
  }

  getByStatus(status: string): Observable<Ambulance[]> {
    return this.http.get<Ambulance[]>(`${this.api}/status/${status}`);
  }

  getAvailableCount(): Observable<number> {
    return this.http.get<number>(`${this.api}/available-count`);
  }
}
